src/main/clojure/clara/tools/inspect.clj

```diff
@@ -121,7 +124,17 @@
                          [rule
                           (for [token (mem/get-tokens-all memory rule-node)
                                 insertion (mem/get-insertions memory rule-node token)]
-                            {:explanation (first (to-explanations session [token])) :fact insertion})]))}))
+                            {:explanation (first (to-explanations session [token])) :fact insertion})]))
+
+     :fact->explanations (reduce (fn [previous [insertion info]]
+                                   (update previous insertion (fnil conj []) info))
+                                 {}
+                                 (mapcat (fn [[rule rule-node]]
+                                           (for [token (mem/get-tokens-all memory rule-node)
+                                                 insertion (mem/get-insertions memory rule-node token)]
+                                             [insertion {:rule rule
+                                                         :explanation (first (to-explanations session [token]))}]))
+                                         rule-to-nodes))}))
```

Comment:
I agree; I think this is cleaner.  It is probably a little slower but I doubt this is significant, particularly for a function that is primarily for debugging rather than routine use (although it is possible that others have a different use case).

I submitted a new pull request with the change at https://github.com/rbrush/clara-rules/pull/163